### PR TITLE
feat(core): Unify application components shutdown

### PR DESCRIPTION
**Pull request by tomi**
## Summary

Add `ShutdownService` and `OnShutdown` decorator for more unified way to shutdown different components. Use this new way in the following components:

- HTTP(S) server
- Pruning service
- Push connection
- License

## Related tickets and issues

https://linear.app/n8n/issue/ADO-1617/unify-application-component-shutdown

## Review / Merge checklist
- [x] PR title and summary are descriptive.
- [x] Tests included.

### CODE REVIEW

**Review by netroy — DISMISSED** [review]
Thanks. This looks great ❤️ 

Ideas for next steps:
1. Move all shutdown logic out of commands, and into specific services
2. Convert `@/Db` into a service class, that handles closing of the DB connection itself
3. Move SIGINT and SIGTERM registration/handling into `ShutdownService`, and add integration tests to make sure that `process.emit('SIGINT')` causes all `OnShutdown` methods to be called in the correct order.
4. Move graceful shutdown timer and `process.exit` into `ShutdownService`
3. Start adding a `OnStartup` decorator to do initialization the same way, and significantly simplify app startup sequence.

**Review by krynble — COMMENTED** [review]
One relatively small thing that was on my to-do list for a while and might easily be addressed in this PR is adding the `@OnShutdown` annotation to the `removeAllTriggerAndPollerBasedWorkflows` method that is located in `packages/cli/src/ActiveWorkflowRunner.ts`.

Currently when n8n is shutting down it still continues to start new executions and start new jobs, even if it's theoretically stopping.

This new approach makes it trivial to fix this issue.

Thanks for the amazing work @tomi

File: packages/cli/src/License.ts [hunk 1]
```diff
@@ -30,6 +31,8 @@ export class License {
 
 	private redisPublisher: RedisServicePubSubPublisher;
 
+	private isShuttingdown = false;
```
Comment (ivov) [inline packages/cli/src/License.ts:34]:
```suggestion
	private isShuttingDown = false;
```

File: packages/cli/src/decorators/OnShutdown.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * class MyClass {
+ *   @OnShutdown()
+ *   async shutdown() {
+ * 	   // Will be called when the app is shutting down
+ *   }
+ * }
+ * ```
+ */
+export const OnShutdown =
+	(priority: number = 100): MethodDecorator =>
```
Comment (ivov) [inline packages/cli/src/decorators/OnShutdown.ts:25]:
```suggestion
	(priority = 100): MethodDecorator =>
```

File: packages/cli/src/shutdown/Shutdown.service.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+export interface ShutdownHandler {
+	serviceClass: ServiceClass;
+	methodName: string;
+}
+
+/** Error reported when a listener fails to shutdown gracefully */
+export class ComponentShutdownError extends ApplicationError {
+	constructor(componentName: string, cause: Error) {
+		super('Failed to shutdown gracefully', {
+			level: 'error',
```
Comment (ivov) [inline packages/cli/src/shutdown/Shutdown.service.ts:18]: I think that's the default?

**Comment by tomi** [comment]
> One relatively small thing that was on my to-do list for a while and might easily be addressed in this PR is adding the `@OnShutdown` annotation to the `removeAllTriggerAndPollerBasedWorkflows` method that is located in `packages/cli/src/ActiveWorkflowRunner.ts`.
> 
> Currently when n8n is shutting down it still continues to start new executions and start new jobs, even if it's theoretically stopping.
> 
> This new approach makes it trivial to fix this issue.
> 
> Thanks for the amazing work @tomi

@krynble Thanks! I added the annotation to `ActiveWorkflowRunner.removeAllTriggerAndPollerBasedWorkflows` method. Like you mentioned, it was a trivial fix.

@netroy Could you re-approve 🙏

File: packages/cli/src/shutdown/Shutdown.service.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+			cause,
+			extra: { component: componentName },
+		});
+	}
+}
+
+/** Service responsible for orchestrating a graceful shutdown of the application */
+@Service()
+export class ShutdownService {
+	readonly handlersByPriority: ShutdownHandler[][] = [];
```
Comment (ivov) [inline packages/cli/src/shutdown/Shutdown.service.ts:28]:
```suggestion
	private readonly handlersByPriority: ShutdownHandler[][] = [];
```

**Review by krynble — DISMISSED** [review]

**Review by ivov — COMMENTED** [review]
Loved this, amazing work!

Edit: Reviewed up to [c3d1144](https://github.com/n8n-io/n8n/pull/8097/commits/c3d114405d5abaf31f3c665e9f2e2decabd25663)

**Review by netroy — DISMISSED** [review]

File: packages/cli/src/shutdown/Shutdown.service.ts (hunk 3 above)
Comment (tomi) [reply packages/cli/src/shutdown/Shutdown.service.ts:18]: Could be. I'd prefer to keep it to make it explicit

File: packages/cli/src/shutdown/Shutdown.service.ts (hunk 4 above)
Comment (tomi) [reply packages/cli/src/shutdown/Shutdown.service.ts:28]: We are checking this in the tests. I don't personally like having fields public just for the sake of testing. I changed it private and added an `@ts-expect-error` to the test. I think it's better that way.

**Review by netroy — APPROVED** [review]

**Comment by Joffcom** [comment]
Got released with [`n8n@1.23.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.23.0)

**Pull request merged**